File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-10.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        std::queue<std::pair<TreeNode*, int>> aPath;
        aPath.push({ root, INT_MIN });
        int totalValues = 0;
        while (!aPath.empty())
        {
            std::pair<TreeNode*, int> topValue = aPath.front();
            aPath.pop();

            if (topValue.first->val >= topValue.second)
            {
                totalValues++;
            }

            if (topValue.first->left != nullptr) aPath.push({ topValue.first->left, std::max(topValue.second, topValue.first->val) }); 
            if (topValue.first->right != nullptr) aPath.push({ topValue.first->right, std::max(topValue.second, topValue.first->val) }); 
        }

        return totalValues;
    }
};

File: Data Structures & Algorithms/balanced-binary-tree/submission-10.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        # If any subtree is unbalanced, propagate -1 up immedietely. 
        def traverse(node):
            if not node: 
                return 0

            left = traverse(node.left)
            if left == -1:
                return -1

            right = traverse(node.right)
            if right == -1:
                return -1
            
            if abs(left - right) > 1:
                return -1

            # Key here. Height is the number of edges on the longest path from a node down to a leaf. 
            return 1 + max(left, right)

        return traverse(root) != -1
